# Preamble: A Brand new Jay

After an eventful season on season 8 of *A Brand New Jay*, the 3 remaining contestants were invited to Jay Stacksby's private island for the last three episodes. When the day of filming the finale came Mr. Stacksby was found with one of his Professional Series 8-inch Chef Knives plunged through his heart! After the initial investigation highlighted that the film crew all lived in a separate house on the other side of the island, it was concluded that only the three contestants were near enough to Stacksby in order to commit a crime. At the scene of the crime, a letter was left. Here are the contents of that letter:

> You may call me heartless, a killer, a monster, a murderer, but I'm still NOTHING compared to the villian that Jay was. This whole contest was a sham, an elaborate plot to shame the contestants and feed Jay's massive, massive ego. SURE you think you know him! You've seen him smiling for the cameras, laughing, joking, telling stories, waving his money around like a prop but off camera he was a sinister beast, a cruel cruel taskmaster, he treated all of us like slaves, like cattle, like animals! Do you remember Lindsay, she was the first to go, he called her such horrible things that she cried all night, keeping up all up, crying, crying, and more crying, he broke her with his words. I miss my former cast members, all of them very much. And we had to live with him, live in his home, live in his power, deal with his crazy demands. AND FOR WHAT! DID YOU KNOW THAT THE PRIZE ISN'T REAL? He never intended to marry one of us! The carrot on the stick was gone, all that was left was stick, he told us last night that we were all a terrible terrible disappointment and none of us would ever amount to anything, and that regardless of who won the contest he would never speak to any of us again! It's definitely the things like this you can feel in your gut how wrong he is! Well I showed him, he got what he deserved all right, I showed him, I showed him the person I am! I wasn't going to be pushed around any longer, and I wasn't going to let him go on pretending that he was some saint when all he was was a sick sick twisted man who deserved every bit of what he got. The fans need to know, Jay Stacksby is a vile amalgamation of all things evil and bad and the world is a better place without him.

Pretty sinister stuff! Luckily, in addition to this bold-faced admission, we have the introduction letters of the three contestants. Maybe there is a way to use this information to determine who the author of this murder letter is?

Myrtle Beech's introduction letter:
> Salutations. My name? Myrtle. Myrtle Beech. I am a woman of simple tastes. I enjoy reading, thinking, and doing my taxes. I entered this competition because I want a serious relationship. I want a commitment. The last man I dated was too whimsical. He wanted to go on dates that had no plan. No end goal. Sometimes we would just end up wandering the streets after dinner. He called it a "walk". A "walk" with no destination. Can you imagine? I like every action I take to have a measurable effect. When I see a movie, I like to walk away with insights that I did not have before. When I take a bike ride, there better be a worthy destination at the end of the bike path. Jay seems frivolous at times. This worries me. However, it is my staunch belief that one does not make and keep money without having a modicum of discipline. As such, I am hopeful. I will now list three things I cannot live without. Water. Emery boards. Dogs. Thank you for the opportunity to introduce myself. I look forward to the competition. 

Lily Trebuchet's introduction letter:
> Hi, I'm Lily Trebuchet from East Egg, Long Island. I love cats, hiking, and curling up under a warm blanket with a book. So they gave this little questionnaire to use for our bios so lets get started. What are some of my least favorite household chores? Dishes, oh yes it's definitely the dishes, I just hate doing them, don't you? Who is your favorite actor and why? Hmm, that's a hard one, but I think recently I'll have to go with Michael B. Jordan, every bit of that man is handsome, HANDSOME! Do you remember seeing him shirtless? I can't believe what he does for the cameras! Okay okay next question, what is your perfect date? Well it starts with a nice dinner at a delicious but small restaurant, you know like one of those places where the owner is in the back and comes out to talk to you and ask you how your meal was. My favorite form of art? Another hard one, but I think I'll have to go with music, music you can feel in your whole body and it is electrifying and best of all, you can dance to it! Okay final question, let's see, What are three things you cannot live without? Well first off, my beautiful, beautiful cat Jerry, he is my heart and spirit animal. Second is pasta, definitely pasta, and the third I think is my family, I love all of them very much and they support me in everything I do. I know Jay Stacksby is a handsome man and all of us want to be the first to walk down the aisle with him, but I think he might truly be the one for me. Okay that's it for the bio, I hope you have fun watching the show! 

Gregg T Fishy's introduction letter:

> A most good day to you all, I am Gregg T Fishy, of the Fishy Enterprise fortune. I am 37 years young, an adventurous spirit and I've never lost my sense of childlike wonder. I do love to be in the backyard gardening and I have the most extraordinary time when I'm fishing. Fishing for what, you might find yourself asking? Why, I happen to always be fishing for compliments of course! I have a stunning pair of radiant blue eyes that will pierce the soul of anyone who dare gaze upon my countenance. I quite enjoy going on long jaunts through garden paths and short walks through greenhouses. I hope that Jay will be as absolutely interesting as he appears on the television, I find that he has some of the most curious tastes in style and humor. When I'm out and about I quite enjoy hearing tales that instill in my heart of hearts the fascination that beguiles my every day life, every fiber of my being scintillates and vascillates with extreme pleasure during one of these charming anecdotes and significantly pleases my beautiful personage. I cannot wait to enjoy being on the television program A Jay To Remember, it certainly seems like a grand time to explore life and love.

# Who Do I Think Dunnit? (My initial guess.)

At the onset of this excercise I am taking a guess as to who I think the killer is. I don't think it is Greg. He uses a lot of big words and referes to himself alot (my, I). The murder didn't do this. Myrtle Beech has a similar self focus. She writes about ideas and things. Lily, uses the word "you" more than the other two. She also is the only contestant who ALLCAPS select words. Of the three she is the most emotionally invested of the bunch. 
        
I wonder if I have been dupped by one of the other two, who used the ALLCAPS to frame Lily. I don't think so. I think it is her: emotion, use of the word "you", and the focus on relationships with others. AND OFCOUSE THE USE OF ALLCAPS. 
         
Is it Lily Trebuchet? We shall see.

## Saving The Different Examples as Variables

First let's create variables to hold the text data in! Save the muder note as a string in a variable called `murder_note`. Save Lily Trebuchet's introduction into `lily_trebuchet_intro`. Save Myrtle Beech's introduction into `myrtle_beech_intro`. Save Gregg T Fishy's introduction into `gregg_t_fishy_intro`.

In [23]:
murder_note = "You may call me heartless, a killer, a monster, a murderer, but I'm still NOTHING compared to the villian that Jay was. This whole contest was a sham, an elaborate plot to shame the contestants and feed Jay's massive, massive ego. SURE you think you know him! You've seen him smiling for the cameras, laughing, joking, telling stories, waving his money around like a prop but off camera he was a sinister beast, a cruel cruel taskmaster, he treated all of us like slaves, like cattle, like animals! Do you remember Lindsay, she was the first to go, he called her such horrible things that she cried all night, keeping up all up, crying, crying, and more crying, he broke her with his words. I miss my former cast members, all of them very much. And we had to live with him, live in his home, live in his power, deal with his crazy demands. AND FOR WHAT! DID YOU KNOW THAT THE PRIZE ISN'T REAL? He never intended to marry one of us! The carrot on the stick was gone, all that was left was stick, he told us last night that we were all a terrible terrible disappointment and none of us would ever amount to anything, and that regardless of who won the contest he would never speak to any of us again! It's definitely the things like this you can feel in your gut how wrong he is! Well I showed him, he got what he deserved all right, I showed him, I showed him the person I am! I wasn't going to be pushed around any longer, and I wasn't going to let him go on pretending that he was some saint when all he was was a sick sick twisted man who deserved every bit of what he got. The fans need to know, Jay Stacksby is a vile amalgamation of all things evil and bad and the world is a better place without him."
lily_trebuchet_intro = "Hi, I'm Lily Trebuchet from East Egg, Long Island. I love cats, hiking, and curling up under a warm blanket with a book. So they gave this little questionnaire to use for our bios so lets get started. What are some of my least favorite household chores? Dishes, oh yes it's definitely the dishes, I just hate doing them, don't you? Who is your favorite actor and why? Hmm, that's a hard one, but I think recently I'll have to go with Michael B. Jordan, every bit of that man is handsome, HANDSOME! Do you remember seeing him shirtless? I can't believe what he does for the cameras! Okay okay next question, what is your perfect date? Well it starts with a nice dinner at a delicious but small restaurant, you know like one of those places where the owner is in the back and comes out to talk to you and ask you how your meal was. My favorite form of art? Another hard one, but I think I'll have to go with music, music you can feel in your whole body and it is electrifying and best of all, you can dance to it! Okay final question, let's see, What are three things you cannot live without? Well first off, my beautiful, beautiful cat Jerry, he is my heart and spirit animal. Second is pasta, definitely pasta, and the third I think is my family, I love all of them very much and they support me in everything I do. I know Jay Stacksby is a handsome man and all of us want to be the first to walk down the aisle with him, but I think he might truly be the one for me. Okay that's it for the bio, I hope you have fun watching the show!"
myrtle_beech_intro = "Salutations. My name? Myrtle. Myrtle Beech. I am a woman of simple tastes. I enjoy reading, thinking, and doing my taxes. I entered this competition because I want a serious relationship. I want a commitment. The last man I dated was too whimsical. He wanted to go on dates that had no plan. No end goal. Sometimes we would just end up wandering the streets after dinner. He called it a ""walk"". A ""walk"" with no destination. Can you imagine? I like every action I take to have a measurable effect. When I see a movie, I like to walk away with insights that I did not have before. When I take a bike ride, there better be a worthy destination at the end of the bike path. Jay seems frivolous at times. This worries me. However, it is my staunch belief that one does not make and keep money without having a modicum of discipline. As such, I am hopeful. I will now list three things I cannot live without. Water. Emery boards. Dogs. Thank you for the opportunity to introduce myself. I look forward to the competition."
gregg_t_fishy_intro = "A most good day to you all, I am Gregg T Fishy, of the Fishy Enterprise fortune. I am 37 years young, an adventurous spirit and I've never lost my sense of childlike wonder. I do love to be in the backyard gardening and I have the most extraordinary time when I'm fishing. Fishing for what, you might find yourself asking? Why, I happen to always be fishing for compliments of course! I have a stunning pair of radiant blue eyes that will pierce the soul of anyone who dare gaze upon my countenance. I quite enjoy going on long jaunts through garden paths and short walks through greenhouses. I hope that Jay will be as absolutely interesting as he appears on the television, I find that he has some of the most curious tastes in style and humor. When I'm out and about I quite enjoy hearing tales that instill in my heart of hearts the fascination that beguiles my every day life, every fiber of my being scintillates and vascillates with extreme pleasure during one of these charming anecdotes and significantly pleases my beautiful personage. I cannot wait to enjoy being on the television program A Jay To Remember, it certainly seems like a grand time to explore life and love."

## The First Indicator: Sentence Length

Perhaps some meaningful data can first be gleaned from these text examples if we measure how long the average sentence length is. Different authors have different patterns of written speech, so this could be very useful in tracking down the killer.

Write a function `get_average_sentence_length` that takes some `text` as an argument. This function should return the average length of a sentence in the text.

Hint (highlight this hint in order to reveal it): 
<font color="white">Use your knowledge of _string methods_ to create a list of all of the sentences in a text, called **sentences_in_text**. 
Further break up each **sentences_in_text** into a list of words and save the _length_ of that list of words to a new list that contains all the sentence lengths, called **sentence_lengths**. Take the average of all of the sentence lengths by adding them all together and dividing by the number of sentences (which should be the same as the length of the **sentence_lengths**).

Remember sentences can end with more than one kind of punctuation, you might find it easiest to use **.replace()** so you only have to split on one punctuation mark. Remember **.replace()** doesn't modify the string itself, it returns a new string!</font>

In [24]:
#I did start off writting a function called get_average_sentence_lenght.
#I decided to take it a step further and create a class to manage the data. 
#I should have read to the next step.

class splitText:
    def __init__(self, text, name):
        self.name = name
        self.text = str(text)
        self.list = text.replace(". ", ".& ").replace("! ", "!& ").replace("? ", "?& ").split("& ")
        self.numSentences = len(self.list)
    
    def __repr__(self):
        return self.name
        
    def sentenceList(self):
        return self.list
    
    def total_len(self):
        total_len = 0
        for sentence in self.list:
            total_len += len(sentence)
        return round(total_len,2)
    
    def get_average_sentence_length(self):
        total_len = 0
        for sentence in self.list:
            total_len += len(sentence)
        return round(total_len/self.numSentences,2)
        

murder_note_split = splitText(murder_note, "Murder Note")
lily_trebuchet_split = splitText(lily_trebuchet_intro, "Lily's Intro")
myrtle_beech_intro_split = splitText(myrtle_beech_intro, "Myrtle's Intro")
greg_fishy_intro_split = splitText(gregg_t_fishy_intro, "Greg's Intro")


print ("AVERAGE SENTENCE LENGTHS")
print(murder_note_split, ": ", murder_note_split.get_average_sentence_length())
print(greg_fishy_intro_split, ": ", greg_fishy_intro_split.get_average_sentence_length())
print(lily_trebuchet_split, ": ", lily_trebuchet_split.get_average_sentence_length())
print(myrtle_beech_intro_split, ": ", myrtle_beech_intro_split.get_average_sentence_length())



AVERAGE SENTENCE LENGTHS
Murder Note :  112.8
Greg's Intro :  117.3
Lily's Intro :  79.79
Myrtle's Intro :  35.18


## Creating The Definition for Our Model

Now that we have a metric we want to save and data that is coupled with that metric, it might be time to create our data type. Let's define a class called `TextSample` with a constructor. The constructor should take two arguments: `text` and `author`. `text` should be saved as `self.raw_text`. Call `get_average_sentence_length` with the raw text and save it to `self.average_sentence_length`. You should save the author of the text as `self.author`.

Additionally, define a string representation for the model. If you print a `TextSample` it should render:
 - The author's name
 - The average sentence length
 
This will be your main class for the problem at hand. All later instruction to update `TextSample` should be done in the code block below. After updating `TextSample`, click on the `Cell` option in the Jupyter Notebook main menu above, then click `Run All` to rerun the cells from top to bottom. If you need to restart your Jupyter Notebook either run the cells below first or move the `TextSample` class definition & instantiation cells to the bottom.

In [25]:
class TextSample:
    def __init__(self, text, author):
        self.author = author
        self.raw_text = str(text)
        self.sentence_list = text.replace(". ", ".& ").replace("! ", "!& ").replace("? ", "?& ").split("& ")
        
        self.numSentences = len(self.sentence_list)
        self.average_sentence_length = self.get_average_sentence_length()
        self.prepared_text = self.prepare_text(self.raw_text)
        self.word_count_frequency = self.build_frequency_table(self.prepared_text)
        self.ngram_list = self.ngram_creator(self.prepared_text)
        self.ngram_frequency = self.build_frequency_table(self.ngram_list)
        
    
    def total_len(self):
        total_len = 0
        for sentence in self.list:
            total_len += len(sentence)
            print (total_len)
        return round(total_len,2)
    
    def get_average_sentence_length(self):
        total_len = 0
        for sentence in self.sentence_list:
            total_len += len(sentence)        
        return round(total_len/self.numSentences,2)
    
    def prepare_text(self, text):
        clean_text = text.lower().replace(",","").replace(".","").replace("!","").replace("?","")
        text_list = clean_text.split()
        return text_list
    
    def build_frequency_table(self, corpus):
    
        frequency_list = [
            corpus.count(word) for word in corpus
        ]
        frequency_data = {key:value for key, value in zip(corpus, frequency_list)}
        sorted_data = sorted(frequency_data.items(), key=lambda kv: kv[1], reverse=True)
        frequency_table = {key:value for key,value in sorted_data}
        return frequency_table
    
    def ngram_creator(self,text_list):
        ngram_list = []
        for i in range(len(text_list)-1):
            ngram_list.append("{word} {next_word}".format(word=text_list[i], next_word=text_list[i+1]))

        return ngram_list
    
    def __repr__(self):
        return_string = "{author}, Average Sentence Length:{asl}.".format(author=self.author, asl=self.average_sentence_length)
        return return_string
    

### Creating our TextSample Instances

Now create a `TextSample` object for each of the samples of text that we have.
 - `murderer_sample` for the murderer's note.
 - `lily_sample` for Lily Trebuchet's note.
 - `myrtle_sample` for Myrtle Beech's note.
 - `gregg_sample` for Gregg T Fishy's note.
 
Print out each one after instantiating them.

In [26]:
murderer_sample = TextSample(murder_note, "Murder")
lily_sample = TextSample(lily_trebuchet_intro, "Trebuchet")
myrtle_sample = TextSample(myrtle_beech_intro, "Beech")
gregg_sample = TextSample(gregg_t_fishy_intro, "Fishy")

print (murderer_sample)
print (gregg_sample)
print (lily_sample)
print (myrtle_sample)

Murder, Average Sentence Length:112.8.
Fishy, Average Sentence Length:117.3.
Trebuchet, Average Sentence Length:79.79.
Beech, Average Sentence Length:35.18.


## Cleaning Our Data

We want to compare the word choice and usage between the samples, but sentences make our text data fairly messy. In order to analyze the different messages fairly, we'll need to remove all the punctuation and uppercase letters from the samples.

Create a function called `prepare_text` that takes a single parameter `text`, makes the text entirely lowercase, removes all the punctuation and returns a list of the words in the text in order.

For example: `"Where did you go, friend? We nearly saw each other."` would become `['where', 'did', 'you', 'go', 'friend', 'we', 'nearly', 'saw', 'each', 'other']`.

In [27]:
#Create a function called prepare_text that takes a single parameter text
#make the text entirely lowercase
#remove all the punctuation
#return a list of the words in the text in order.



def prepare_text(text):
    clean_text = text.lower().replace(",","").replace(".","").replace("!","").replace("?","")
    text_list = clean_text.split()
    return text_list



Update the constructor for `TextSample` to save the prepared text as `self.prepared_text`.



In [28]:
#print (murderer_sample.prepared_text)

## Building A Frequency Table

Now we want to see which words were most frequently used in each of the samples. Create a function called `build_frequency_table`. It takes in a list called `corpus` and creates a dictionary called `frequency_table`. For every element in `corpus` the value `frequency_table[element]` should be equal to the number of times that element appears in `corpus`. For example the input `['do', 'you', 'see', 'what', 'i', 'see']` would create the frequency table `{'what': 1, 'you': 1, 'see' 2, 'i': 1}`.

In [29]:
#Create a function called `build_frequency_table`. 
#It takes in a list called `corpus` and creates a dictionary called `frequency_table`. 
#For every element in `corpus` the value `frequency_table[element]` should be equal to the number of times that element appears in `corpus`. 
#For example the input `['do', 'you', 'see', 'what', 'i', 'see']` would create the frequency table `{'what': 1, 'you': 1, 'see' 2, 'i': 1}`.


def build_frequency_table(corpus):
    
    frequency_list = [
        corpus.count(word) for word in corpus
    ]
        
    frequency_data = {key:value for key, value in zip(corpus, frequency_list)}
    
    sorted_data = sorted(frequency_data.items(), key=lambda kv: kv[1], reverse=True)
    
    frequency_table = {key:value for key,value in sorted_data}
    
    return frequency_table

#print (build_frequency_table(murderer_sample.prepared_text))


## The Second Indicator: Favorite Words

Use `build_frequency_table` with the prepared text to create a frequency table that counts how frequently all the words in each text sample appears. Call these functions in the constructor for `TextSample` and assign the word frequency table to a value called `self.word_count_frequency`.

In [30]:
#Call these functions in the constructor for TextSample
#assign the word frequency table to a value called self.word_count_frequency.
#Use build_frequency_table with the prepared text to create a frequency table
#count how frequently all the words in each text sample appears.

print (murderer_sample)
print (murderer_sample.word_count_frequency)
print (" ")

print (lily_sample)
print (lily_sample.word_count_frequency)
print (" ")

print (myrtle_sample)
print (myrtle_sample.word_count_frequency)
print (" ")

print (gregg_sample)
print (gregg_sample.word_count_frequency)
print (" ")

Murder, Average Sentence Length:112.8.
{'he': 13, 'the': 12, 'a': 11, 'to': 10, 'was': 10, 'and': 9, 'all': 9, 'him': 8, 'of': 8, 'that': 7, 'i': 7, 'you': 6, 'his': 5, 'like': 5, 'us': 5, 'know': 3, 'things': 3, 'crying': 3, 'with': 3, 'live': 3, 'in': 3, 'what': 3, 'is': 3, 'showed': 3, 'but': 2, 'jay': 2, 'this': 2, 'contest': 2, 'massive': 2, 'for': 2, 'around': 2, 'cruel': 2, 'she': 2, 'go': 2, 'her': 2, 'night': 2, 'up': 2, 'we': 2, 'never': 2, 'on': 2, 'stick': 2, 'terrible': 2, 'would': 2, 'who': 2, 'any': 2, 'got': 2, 'deserved': 2, "wasn't": 2, 'going': 2, 'sick': 2, 'may': 1, 'call': 1, 'me': 1, 'heartless': 1, 'killer': 1, 'monster': 1, 'murderer': 1, "i'm": 1, 'still': 1, 'nothing': 1, 'compared': 1, 'villian': 1, 'whole': 1, 'sham': 1, 'an': 1, 'elaborate': 1, 'plot': 1, 'shame': 1, 'contestants': 1, 'feed': 1, "jay's": 1, 'ego': 1, 'sure': 1, 'think': 1, "you've": 1, 'seen': 1, 'smiling': 1, 'cameras': 1, 'laughing': 1, 'joking': 1, 'telling': 1, 'stories': 1, 'waving': 

## The Third Indicator: N-Grams

An <a href='https://en.wikipedia.org/wiki/N-gram' target="_blank">n-gram</a> is a text analysis technique used for pattern recognition and applicable throughout lingusitics. We're going to use n-grams to find who uses similar word-pairs to the murderer, and we think it's going to make our evidence strong enough to conclusively find the killer.

Create a function called `ngram_creator` that takes a parameter `text_list`, a treated in-order list of the words in a text sample. `ngram_creator` should return a list of all adjacent pairs of words, styled as strings with a space in the center.

For instance, calling `ngram_creator` with the input `['what', 'in', 'the', 'world', 'is', 'going', 'on']`
Should produce the output `['what in', 'in the', 'the world', 'world is', 'is going', 'going on']`.

These are two-word n-grams.

In [31]:
#Create a function called ngram_creator-
#that takes a parameter text_list, a treated in-order list of the words in a text sample (prepared_text) -
#ngram_creator should return a list of all adjacent pairs of words, styled as strings with a space in the center.-

def ngram_creator(text_list):
    ngram_list = []
    for i in range(len(text_list)-1):
        ngram_list.append("{word} {next_word}".format(word=text_list[i], next_word=text_list[i+1]))
    
    return ngram_list

#print (ngram_creator(murderer_sample.prepared_text))


Use `ngram_creator` along with the prepared text to create a list of all the two-word ngrams in each `TextSample`. Use `build_frequency_table` to tabulate the frequency of each ngram. In the constructor for `TextSample` save this frequency table as `self.ngram_frequency`.

In [32]:
print("Murderer's Sample")
print (murderer_sample.ngram_frequency)
print ("")
print("Lily's Sample")
print (lily_sample.ngram_frequency)
print ("")
print("Gregg's Sample")
print (gregg_sample.ngram_frequency)
print ("")
print("Myrtle's Sample")

print (myrtle_sample.ngram_frequency)

Murderer's Sample
{'of us': 4, 'was a': 3, 'he was': 3, 'i showed': 3, 'showed him': 3, 'you know': 2, 'all of': 2, 'with his': 2, 'live in': 2, 'in his': 2, 'he got': 2, 'what he': 2, "i wasn't": 2, "wasn't going": 2, 'going to': 2, 'is a': 2, 'you may': 1, 'may call': 1, 'call me': 1, 'me heartless': 1, 'heartless a': 1, 'a killer': 1, 'killer a': 1, 'a monster': 1, 'monster a': 1, 'a murderer': 1, 'murderer but': 1, "but i'm": 1, "i'm still": 1, 'still nothing': 1, 'nothing compared': 1, 'compared to': 1, 'to the': 1, 'the villian': 1, 'villian that': 1, 'that jay': 1, 'jay was': 1, 'was this': 1, 'this whole': 1, 'whole contest': 1, 'contest was': 1, 'a sham': 1, 'sham an': 1, 'an elaborate': 1, 'elaborate plot': 1, 'plot to': 1, 'to shame': 1, 'shame the': 1, 'the contestants': 1, 'contestants and': 1, 'and feed': 1, "feed jay's": 1, "jay's massive": 1, 'massive massive': 1, 'massive ego': 1, 'ego sure': 1, 'sure you': 1, 'you think': 1, 'think you': 1, 'know him': 1, "him you've"

## Comparing Two Frequency Tables

We want to know how similar two frequency tables are, let's write a function that computes the comparison between two frequency tables and scores them based on similarity.

Write a function called `frequency_comparison` that takes two parameters, `table1` and `table2`. It should define two local variables, `appearances` and `mutual_appearances`. 

Iterate through `table1`'s keys and check if `table2` has the same key defined. If it is, compare the two values for the key -- the smaller value should get added to `mutual_appearances` and the larger should get added to `appearances`. If the key doesn't exist in `table2` the value for the key in `table1` should be added to `appearances`.

Remember afterwards to iterate through all of `table2`'s keys that aren't in `table1` and add those to `appearances` as well.

Return a frequency comparison score equal to the mutual appearances divided by the total appearances.

In [33]:
#Write a function called frequency_comparison that takes two parameters, table1 and table2. 

def frequency_comparison(table1, table2): #word_count_frequency ['test':4, 'bold':3]
    
#It should define two local variables, appearances and mutual_appearances.
    appearances = 0
    mutual_appearances = 0

#Iterate through table1's keys and check if table2 has the same key defined. 

    for key,value in table1.items():
        #If it is, compare the two values for the key 
        if key in table2.keys():
            #the smaller value should get added to mutual_appearance 
            mutual_appearances += min(table1.get(key), table2.get(key))
            
            #the larger should get added to appearances.
            appearances += max(table1.get(key), table2.get(key))
             
        #If the key doesn't exist in table2 the value for the key in table1 should be added to appearances.
        else:
            appearances += table1.get(key)

    #iterate through all of table2's keys
    for key,value in table2.items():
        #that aren't in table1 
        if key not in table1.keys():
            #add those to appearances as well.
            appearances += table2.get(key)
            
            
    #Return a frequency comparison score equal to the mutual appearances divided by the total appearances.
    return round(mutual_appearances/appearances, 2)


print ("% Simiilarity: Lily Trebuchet")
print (frequency_comparison(murderer_sample.word_count_frequency,lily_sample.word_count_frequency))
print (" ")

print ("% Simiilarity: Myrtle Beech")
print (frequency_comparison(murderer_sample.word_count_frequency,myrtle_sample.word_count_frequency))
print (" ")

print ("% Simiilarity: Gregg Fishy")
print (frequency_comparison(murderer_sample.word_count_frequency,gregg_sample.word_count_frequency))
print (" ")



% Simiilarity: Lily Trebuchet
0.23
 
% Simiilarity: Myrtle Beech
0.17
 
% Simiilarity: Gregg Fishy
0.16
 


In [34]:
# This is an additional step I took.
# I want to generate a list of the mutual ngrams shared by the murderer and each of the contestants.
# I want to see what the mutual ngrams were before giving them more consideration in my analysis.
# you can see the result in the results area.

def frequent_ngram_list(table1, table2):
    mutual_ngrams = []

    for key,value in table1.items():
        if key in table2.keys():
            mutual_ngrams.append(key)           
    return mutual_ngrams



## Comparing Average Sentence Length

In order to calculate the change between the average sentence lengths of two `TextSamples` we're going to use the formula for the percent difference.

Write a function called `percent_difference` that returns the percent difference as calculated from the following formula:

$$\frac{|\ value1 - value2\ |}{\frac{value1 + value2}{2}}$$

In the numerator is the absolute value (use `abs()`) of the two values subtracted from each other. In the denominator is the average of the two values (value1 + value2 divided by two).

In [35]:
def percent_difference(value1, value2):
    return round(abs(value1 - value2)/((value1 + value2)/2),2)

    


## Scoring Similarity with All Three Indicators

We want to figure out who did it, so let's use all three of the indicators we built to score text similarity. Define a function `find_text_similarity` that takes two `TextSample` arguments and returns a float between 0 and 1 where 0 means completely different and 1 means the same exact sample. You can evaluate the similarity by the following criteria:

- Calculate the percent difference of their average sentence length using `percent_difference`. Save that into a variable called `sentence_length_difference`. Since we want to find how _similar_ the two passages are calculate the inverse of `sentence_length_difference` by using the formula `abs(1 - sentence_length_difference)`. Save that into a variable called `sentence_length_similarity`.
- Calculate the difference between their word usage using `frequency_comparison` on both `TextSample`'s `word_count_frequency` attributes. Save that into a variable called `word_count_similarity`.
- Calculate the difference between their two-word ngram using `frequency_table` on both `TextSample`'s `ngram_frequency` attributes. Save that into a variable called `ngram_similarity`.
- Add all three similarities together and divide by 3.

In [36]:
def find_text_similarity(textsample1, textsample2):
    
    
    sentence_length_difference = percent_difference(textsample1.average_sentence_length, textsample2.average_sentence_length)
    sentence_length_similarity = round(abs(1-sentence_length_difference),2)
    
    word_count_similarity = frequency_comparison(textsample1.word_count_frequency,textsample2.word_count_frequency)
    
    ngram_similarity = frequency_comparison(textsample1.ngram_frequency,textsample2.ngram_frequency)
    
    average_similarity = round((sentence_length_similarity + word_count_similarity + ngram_similarity)/3, 2)
    returnString = "{sls}\t\t\t{wcs}\t\t\t{ngs}\t\t{avs}.".format(author=textsample2.author, sls=sentence_length_similarity, wcs=word_count_similarity, ngs=ngram_similarity, avs=average_similarity)
    return returnString


## Rendering the Results

We want to print out the results in a way that we can read! For each contestant on _A Brand New Jay_ print out the following:

- Their name
- Their similarity score to the murder letter

In [37]:
print ("\n")
print ("THE RESULTS:\n")

print ("NAME\t\tSENTENCE-LENGTH\t\tWORD-COUNT\t\tNGRAM\t\tAVERAGE")
print (murderer_sample.author,"\t\t", find_text_similarity(murderer_sample, murderer_sample))
print (lily_sample.author,"\t", find_text_similarity(murderer_sample, lily_sample))
print (gregg_sample.author,"\t\t", find_text_similarity(murderer_sample, gregg_sample))
print (myrtle_sample.author,"\t\t", find_text_similarity(murderer_sample, myrtle_sample))

print ("")

print ("Common Ngrams:\n")
print ("Lily Trebuchet")
print (frequent_ngram_list(murderer_sample.ngram_frequency,lily_sample.ngram_frequency))
print (" ")

print ("Gregg Fishy")
print (frequent_ngram_list(murderer_sample.ngram_frequency,gregg_sample.ngram_frequency))
print (" ")

print ("Myrtle Beech")
print (frequent_ngram_list(murderer_sample.ngram_frequency,myrtle_sample.ngram_frequency))
print (" ")

print ("")
print ("Favorite Words:\n")
print ("Lily Trebuchet")
print (frequent_ngram_list(murderer_sample.word_count_frequency,lily_sample.word_count_frequency))
print (" ")

print ("Gregg Fishy")
print (frequent_ngram_list(murderer_sample.word_count_frequency,gregg_sample.word_count_frequency))
print (" ")

print ("Myrtle Beech")
print (frequent_ngram_list(murderer_sample.word_count_frequency,myrtle_sample.word_count_frequency))
print (" ")



THE RESULTS:

NAME		SENTENCE-LENGTH		WORD-COUNT		NGRAM		AVERAGE
Murder 		 1.0			1.0			1.0		1.0.
Trebuchet 	 0.66			0.23			0.06		0.32.
Fishy 		 0.96			0.16			0.02		0.38.
Beech 		 0.05			0.17			0.01		0.08.

Common Ngrams:

Lily Trebuchet
['of us', 'you know', 'all of', 'what he', 'is a', 'for the', 'the cameras', 'do you', 'you remember', 'the first', 'first to', 'to go', 'of them', 'them very', 'very much', 'much and', 'with him', 'one of', "it's definitely", 'definitely the', 'you can', 'can feel', 'feel in', 'in your', 'he is', 'to be', 'every bit', 'bit of', 'know jay', 'jay stacksby', 'stacksby is', 'of all', 'and the']
 
Gregg Fishy
['that jay', 'like a', 'for what', 'one of', 'on the', 'i am', 'to be', 'and i', 'that he']
 
Myrtle Beech
['to the', 'for the', 'to go', 'he called', 'i am', 'go on']
 

Favorite Words:

Lily Trebuchet
['he', 'the', 'a', 'to', 'was', 'and', 'all', 'him', 'of', 'that', 'i', 'you', 'like', 'us', 'know', 'things', 'with', 'live', 'in', 'what', 'is', 'bu

There are some problems with looking at the average of the three scores. It is easy to look at the average scores and determine that Fishy has the highest Average Similarity. There are some issues with this. His score of .38 is only slightly higher than Trebuchet's Average Score of .32.

Fishy's high average score is a result of his really high sentence length score. As a group the sentence length scores are big numbers relative to the other scores. Taking the average gives more weight to these big numbers. The average favors setence length. While Ngram Similarity & Word Count Similarity are relativly smaller values, the measurement is more informative than sentence length. 

Trebuchet actually scored higher in two of the three categories: Ngram Similarity, & Word Count Similarity. Significantly higher. Her common bigrams triple Fishy's. She She has almost 1.5 times as many common favorite words as he does. Trebuchet has a sentence length similarity of 66%. That's pretty high.

# Who Dunnit?

In the cell below, print the name of the person who killed Jay Stacksby.

Lily Trebuchet